libslide/src/parser.rs

```diff
@@ -0,0 +1,138 @@
+mod types;
+pub use types::*;
+use crate::scanner::{Token, TokenType};
+
+
+pub struct Parser<'a> {
+    input: &'a Vec<Token>,
+    index: usize,
+    cur_token: Token,
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(input: &Vec<Token>) -> Parser {
+        Parser {
+            input: input.into(),
+            index: 1,
+            cur_token: input[0].clone().into(),
+        }
+    }
+
+    fn get_token(&mut self) -> Token {
+        self.index = self.index + 1;
+        return self.input[self.index - 1].clone();
+    }
+
+    pub fn expr(&mut self) -> Box<Expr> {
+        return self.add_sub_term();
+    }
+
+    fn add_sub_term(&mut self) -> Box<Expr> {
+        let mut node = self.mul_divide_mod_term();
+        if self.cur_token.token_type == TokenType::Plus
+            || self.cur_token.token_type == TokenType::Minus
+        {
+            let t = self.cur_token.clone();
+            self.cur_token = self.get_token();
+            node = Box::new(Expr::BinOp(BinOp {
+                item: t,
+                lhs: node,
+                rhs: self.mul_divide_mod_term(),
+            }));
+        }
```

Comment:
to avoid a mut reference, we could have a match expression here return the value.